### ISSUE TITLE
精度不对

### ISSUE THREAD

**Issue by Truth-Ke** [DeepLink-org/DLCompiler#106]
跑下面例子，精度跟ascend无法对齐。
## DLC: 
dlcompiler) root@PJ910000110042:/mnt/data01/kezengxiang/work/bak/DLCompiler# python .vscode/kenel-llm.py
Using device: npu
WARNING: Grid 12544 > physical limit 48, performance maybe reduced.
Validation results:
  output shape: (4, 32, 158, 158)
  max abs diff: 1.057489e+00
  mean abs diff: 7.458288e-02
  mse: 1.436323e-02
  allclose(rtol=0.0001, atol=1e-05): False
  -> NOTE: not within tolerances. Consider checking kernel logic (weight orientation, padding signs, stride/dilation handling).
## acsend:
(ascend) root@PJ910000110042:/mnt/data01/kezengxiang/work/triton-ascend# python .vscode/.dump/kenel-llm.py 
Using device: npu
WARNING: Grid 12544 > physical limit 48, performance maybe reduced.
Validation results:
  output shape: (4, 32, 158, 158)
  max abs diff: 1.542643e-04
  mean abs diff: 7.525331e-06
  mse: 2.577082e-10
  allclose(rtol=0.0001, atol=1e-05): False
  -> NOTE: not within tolerances. Consider checking kernel logic (weight orientation, padding signs, stride/dilation handling).
(ascend) root@PJ910000110042:/mnt/data01/kezengxiang/work/triton-ascend# 

## case

```
import torch
import torch_npu
import torch.nn as nn
import triton
import triton.language as tl
import math


@triton.jit
def _conv_transpose2d_kernel(
    x_ptr,
    w_ptr,
    output_ptr,
    bias_ptr,
    stride_h,
    stride_w,
    padding_h,
    padding_w,
    dilation_h,
    dilation_w,
    in_channels,
    out_channels,
    kernel_h,
    kernel_w,
    h_in,
    w_in,
    h_out,
    w_out,
    stride_xn,
    stride_xc,
    stride_xh,
    stride_xw,
    stride_wn,
    stride_wc,
    stride_wh,
    stride_ww,
    stride_on,
    stride_oc,
    stride_oh,
    stride_ow,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(0)
    num_pid_n = tl.cdiv(h_out * w_out, BLOCK_SIZE)
    pid_b = pid // (num_pid_n * out_channels)
    pid_c = (pid // num_pid_n) % out_channels
    pid_n = pid % num_pid_n

    offs_n = pid_n * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
    offs_h = offs_n // w_out
    offs_w = offs_n % w_out
    mask = offs_n < (h_out * w_out)

    if bias_ptr is not None:
        bias = tl.load(bias_ptr + pid_c)
    else:
        bias = 0.0
    acc = tl.zeros((BLOCK_SIZE,), dtype=tl.float32) + bias

    x_offset = pid_b * stride_xn
    w_offset = pid_c * stride_wc

    for c_in in range(in_channels):
        w_c_offset = w_offset + c_in * stride_wn
        for kh in range(kernel_h):
            for kw in range(kernel_w):
                h_in_val = offs_h + padding_h - kh * dilation_h
                w_in_val = offs_w + padding_w - kw * dilation_w

                cond_h = (h_in_val >= 0) & (h_in_val < h_in * stride_h)
                cond_w = (w_in_val >= 0) & (w_in_val < w_in * stride_w)
                cond = cond_h & cond_w & mask

                cond_h_stride = h_in_val % stride_h == 0
                cond_w_stride = w_in_val % stride_w == 0
                cond_stride = cond_h_stride & cond_w_stride

                final_cond = cond & cond_stride

                h_in_idx = tl.where(final_cond, h_in_val // stride_h, 0)
                w_in_idx = tl.where(final_cond, w_in_val // stride_w, 0)

                x_offsets = (
                    x_offset
                    + c_in * stride_xc
                    + h_in_idx * stride_xh
                    + w_in_idx * stride_xw
                )
                w_val = tl.load(w_ptr + w_c_offset + kh * stride_wh + kw * stride_ww)
                x_vals = tl.load(x_ptr + x_offsets, mask=final_cond, other=0.0)

                acc += x_vals * w_val

    out_offset = (
        pid_b * stride_on + pid_c * stride_oc + offs_h * stride_oh + offs_w * stride_ow
    )
    tl.store(output_ptr + out_offset, acc, mask=mask)


class ModelNew(nn.Module):
    def __init__(
        self,
        in_channels: int,
        out_channels: int,
        kernel_size: int,
        stride: int = 1,
        padding: int = 0,
        dilation: int = 1,
        bias: bool = False,
    ):
        super(ModelNew, self).__init__()
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding
        self.dilation = dilation

        self.weight = nn.Parameter(
            torch.empty(in_channels, out_channels, kernel_size, kernel_size)
        )

        if bias:
            self.bias = nn.Parameter(torch.empty(out_channels))
        else:
            self.register_parameter("bias", None)

        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))
        if self.bias is not None:
            fan_in, _ = nn.init._calculate_fan_in_and_fan_out(self.weight)
            bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
            nn.init.uniform_(self.bias, -bound, bound)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        batch, in_channels, h_in, w_in = x.shape
        h_out = (
            (h_in - 1) * self.stride
            - 2 * self.padding
            + self.dilation * (self.kernel_size - 1)
            + 1
        )
        w_out = (
            (w_in - 1) * self.stride
            - 2 * self.padding
            + self.dilation * (self.kernel_size - 1)
            + 1
        )

        output = torch.empty(
            (batch, self.out_channels, h_out, w_out), device=x.device, dtype=x.dtype
        )

        # Precompute strides
        stride_xn, stride_xc, stride_xh, stride_xw = x.stride()
        stride_wn, stride_wc, stride_wh, stride_ww = self.weight.stride()
        stride_on, stride_oc, stride_oh, stride_ow = output.stride()

        total_blocks = batch * self.out_channels * math.ceil((h_out * w_out) / 256)
        grid = (total_blocks,)

        bias_ptr = self.bias if self.bias is not None else None

        _conv_transpose2d_kernel[grid](
            x,  # x_ptr
            self.weight,  # w_ptr
            output,  # output_ptr
            bias_ptr,  # bias_ptr (only once)
            self.stride,  # stride_h
            self.stride,  # stride_w
            self.padding,  # padding_h
            self.padding,  # padding_w
            self.dilation,  # dilation_h
            self.dilation,  # dilation_w
            self.in_channels,
            self.out_channels,
            self.kernel_size,
            self.kernel_size,
            h_in,
            w_in,
            h_out,
            w_out,
            stride_xn,
            stride_xc,
            stride_xh,
            stride_xw,
            stride_wn,
            stride_wc,
            stride_wh,
            stride_ww,
            stride_on,
            stride_oc,
            stride_oh,
            stride_ow,
            256,
        )
        return output


# Test code
batch_size = 16
in_channels = 32
out_channels = 32
kernel_size = 3
height_in = 32
width_in = 128
stride = 5
padding = 1
dilation = 2


def get_inputs():
    x = torch.randn(batch_size, in_channels, height_in, width_in)
    return [x]


def get_init_inputs():
    return [in_channels, out_channels, kernel_size, stride, padding, dilation]




def choose_device():
    # try npu
    try:
        dev = torch.device("npu")
        # try a tiny tensor move to check availability
        torch.empty(1).to(dev)
        return dev
    except Exception:
        pass
    if torch.cuda.is_available():
        return torch.device("cuda")
    return torch.device("cpu")


# -------------------------
# 验证函数（前向数值对比）
# -------------------------
def validate_forward(model: nn.Module, 
                     in_tensor: torch.Tensor, 
                     stride: int, padding: int, dilation: int,
                     rtol=1e-4, atol=1e-5):
    """
    model: 你的 ModelNew 实例（参数已初始化）
    in_tensor: 输入 tensor（未必在正确 device）
    stride/padding/dilation: 用于创建 reference ConvTranspose2d
    返回：打印并返回字典结果
    """
    device = in_tensor.device
    # build reference PyTorch conv_transpose2d with same params
    conv_ref = nn.ConvTranspose2d(
        in_channels=model.in_channels,
        out_channels=model.out_channels,
        kernel_size=model.kernel_size,
        stride=stride,
        padding=padding,
        dilation=dilation,
        bias=(model.bias is not None),
    ).to(device).to(in_tensor.dtype)

    # copy weights & bias (注意 shape: ModelNew.weight is [in, out, kh, kw] 同 PyTorch ConvTranspose2d)
    with torch.no_grad():
        conv_ref.weight.data.copy_(model.weight.data)
        if model.bias is not None:
            conv_ref.bias.data.copy_(model.bias.data)

    # forward both
    model.eval()
    conv_ref.eval()
    with torch.no_grad():
        out_custom = model(in_tensor)
        out_ref = conv_ref(in_tensor)

    # compare
    diff = out_custom - out_ref
    max_abs = diff.abs().max().item()
    mean_abs = diff.abs().mean().item()
    mse = (diff ** 2).mean().item()
    allclose = torch.allclose(out_custom, out_ref, rtol=rtol, atol=atol)

    result = {
        "out_shape": tuple(out_custom.shape),
        "max_abs": max_abs,
        "mean_abs": mean_abs,
        "mse": mse,
        "allclose": bool(allclose),
        "rtol": rtol,
        "atol": atol,
    }

    print("Validation results:")
    print(f"  output shape: {result['out_shape']}")
    print(f"  max abs diff: {max_abs:.6e}")
    print(f"  mean abs diff: {mean_abs:.6e}")
    print(f"  mse: {mse:.6e}")
    print(f"  allclose(rtol={rtol}, atol={atol}): {result['allclose']}")
    if not result["allclose"]:
        print("  -> NOTE: not within tolerances. Consider checking kernel logic (weight orientation, padding signs, stride/dilation handling).")
    else:
        print("  -> PASS: forward outputs match within tolerances.")
    return result


def main_quick_test():
    device = choose_device()
    print("Using device:", device)

    batch_size = 4
    in_channels = 32
    out_channels = 32
    kernel_size = 3
    height_in = 32
    width_in = 32
    stride = 5
    padding = 1
    dilation = 2

    torch.manual_seed(0)

    # 构造模型（使用你给的 ModelNew）
    model = ModelNew(in_channels, out_channels, kernel_size, stride=stride, padding=padding, dilation=dilation, bias=True)
    model.to(device)
    # 确保参数在设备上并 contiguous
    model.weight.data = model.weight.data.to(device).contiguous()
    if model.bias is not None:
        model.bias.data = model.bias.data.to(device).contiguous()

    # 随机输入
    x = torch.randn(batch_size, in_channels, height_in, width_in, device=device, dtype=model.weight.dtype).contiguous()

    # 运行验证
    validate_forward(model, x, stride=stride, padding=padding, dilation=dilation, rtol=1e-4, atol=1e-5)


if __name__ == "__main__":
    main_quick_test()

```
